components/automate-ui/src/app/modules/infra-proxy/infra-role-details/infra-role-details.component.ts

```diff
@@ -0,0 +1,216 @@
+      this.store.select(routeParams).pipe(pluck('name'), filter(identity))
+    ]).pipe(
+      takeUntil(this.isDestroyed)
+    ).subscribe(([server_id, org_id, name]: string[]) => {
+      this.serverId = server_id;
+      this.OrgId = org_id;
+      this.name = name;
+      this.store.dispatch(new GetRole({
+        server_id: server_id, org_id: org_id, name: name
+      }));
+    });
+
+    this.store.select(infaRoleFromRoute).pipe(
+      filter(identity),
+      takeUntil(this.isDestroyed)
+    ).subscribe(role => {
+      this.show = true;
+      this.role = role;
+      this.expandedList = role.expanded_run_list;
+      this.runList = this.role.run_list;
+      this.idList = [];
+      this.attributes = new RoleAttributes(this.role);
+      if (this.expandedList && this.expandedList.length) {
+        this.show = true;
+        for (let i = 0; i < this.expandedList.length; i++) {
+          this.idList.push(
+            this.expandedList[i].id
+          );
+        }
+        if (this.tabValue === 'runList') {
+          this.treeNodes(this.expandedList, this.env_id);
+        }
+
+      } else {
+        this.show = false;
+      }
+      setTimeout(() => this.filter(this.selected_level), 10);
+      this.roleDetailsLoading = false;
+    });
+  }
+
+  // set selected item to selected_level
+  filter(precedence_level: string): void {
+    this.selected_level = precedence_level;
+    this.selectedAttrs = this.retrieve(precedence_level);
+  }
+
+  // retrieve attributes based on their level of precedence
+  retrieve(level: string): any {
+    switch (level) {
+      case this.all: {
+        return this.attributes.all;
+      }
+      case this.default_attributes: {
+        return this.attributes.default_attributes;
+      }
+      case this.override_attributes: {
+        return this.attributes.override_attributes;
+      }
+      default: {
+        return {};
+      }
+    }
+  }
+
+  selectChangeHandler(id) {
+    this.env_id = id;
+    this.treeNodes(this.expandedList, this.env_id);
+  }
+
+  treeNodes(expandedList: ExpandedList[], li: string) {
+    this.arrayOfNodesTree = [];
+    for (let i = -0; i < expandedList.length; i++) {
+      if (expandedList[i].id === li) {
+        this.expandRunList = expandedList[i].run_list;
+        if (this.expandRunList && this.expandRunList.length) {
+          this.hasRun_List = true;
+          for (let j = 0; j < this.expandRunList.length; j++) {
+            const nodes: Node<ChildList>[] = [];
+            this.arrayOfNodesTree.push({
+              value: {
+                name: this.expandRunList[j].name,
+                version: this.expandRunList[j].version === ''
+                  ? '...' : this.expandRunList[j].version,
+                type: this.expandRunList[j].type
+              },
+              children:
+                this.expandRunList[j].children && this.expandRunList[j].children.length ?
+                  this.childNode(this.expandRunList[j].children, nodes) : []
```

Comment:
What is `nodes` supposed to be doing? It is always an empty array.
```suggestion
                  this.childNode(this.expandRunList[j].children, []]) : []
```